core/src/main/java/brooklyn/entity/rebind/RebindIteration.java

```diff
@@ -341,38 +340,67 @@ protected void rebuildCatalog() {
+                
+                PersisterDeltaImpl delta = new PersisterDeltaImpl();
+                for (String catalogItemId: mementoRawData.getCatalogItems().keySet()) {
+                    delta.removedCatalogItemIds.add(catalogItemId);
+                }
+                getPersister().queueDelta(delta);
+                
+                mementoRawData.clearCatalogItems();
+                needsInitialCatalog = true;
             } else {
-                // Management context should have taken care of loading the catalogue
-                Collection<CatalogItem<?, ?>> catalogItems = rebindContext.getCatalogItems();
-                String message = "RebindManager not resetting catalog to persisted state. Catalog load mode is {}.";
-                if (!catalogItems.isEmpty() && shouldLogRebinding()) {
-                    LOG.info(message + " There {} {} item{} persisted.", new Object[]{
-                            catalogLoadMode, catalogItems.size() == 1 ? "was" : "were", catalogItems.size(), Strings.s(catalogItems)});
-                } else if (LOG.isDebugEnabled()) {
-                    logRebindingDebug(message, catalogLoadMode);
+                if (!isEmpty) {
+                    logRebindingDebug("RebindManager clearing local catalog and loading from persisted state");
+                    itemsForResettingCatalog = rebindContext.getCatalogItems();
+                    needsInitialCatalog = false;
+                } else {
+                    if (catInit.hasRunOfficial()) {
+                        logRebindingDebug("RebindManager will re-add any new items (persisted state empty)");
+                        needsInitialCatalog = false;
+                    } else {
+                        logRebindingDebug("RebindManager loading initial catalog locally because persisted state empty");
+                        needsInitialCatalog = true;
+                    }
                 }
             }
-            // TODO destroy old (as above)
         } else {
-            logRebindingDebug("RebindManager not resetting catalog because catalog persistence is disabled");
+            if (catInit.hasRunOfficial()) {
+                logRebindingDebug("RebindManager skipping catalog init because it has already run (catalog persistence disabled)");
+                needsInitialCatalog = false;
+            } else {
+                logRebindingDebug("RebindManager will initialize catalog locally because catalog persistence is disabled");
+                needsInitialCatalog = true;
+            }
         }
+
+        // TODO in read-only mode, perhaps do this less frequently than entities etc ?
+        // both in RW and in RO mode, the first run reads the initialization data;
+        // maybe not desired for RO as it defers problems, although if it's standalone it is desired
+        catInit.populateCatalog(needsInitialCatalog, itemsForResettingCatalog);
```

Comment:
Personal preference: I like `Optional<Collection<CatalogItem<?,?>>>` rather than using null to indicate absence. That makes the behaviour of the method more explicit.